src/main/flight/interpolated_setpoint.c

```diff
@@ -0,0 +1,125 @@
+/*
+ * This file is part of Cleanflight and Betaflight.
+ *
+ * Cleanflight and Betaflight are free software. You can redistribute
+ * this software and/or modify this software under the terms of the
+ * GNU General Public License as published by the Free Software
+ * Foundation, either version 3 of the License, or (at your option)
+ * any later version.
+ *
+ * Cleanflight and Betaflight are distributed in the hope that they
+ * will be useful, but WITHOUT ANY WARRANTY; without even the implied
+ * warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
+ * See the GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this software.
+ *
+ * If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#include "interpolated_setpoint.h"
+#include "build/debug.h"
+#include "fc/rc.h"
+#include "common/maths.h"
+#include "math.h"
+
+#ifdef USE_INTERPOLATED_SP
+static FAST_RAM_ZERO_INIT float projectedSetpoint[XYZ_AXIS_COUNT];
+static FAST_RAM_ZERO_INIT float prevRawSetpoint[XYZ_AXIS_COUNT];
+static FAST_RAM_ZERO_INIT float prevRawDeflection[XYZ_AXIS_COUNT];
+static FAST_RAM_ZERO_INIT uint16_t interpolationSteps[XYZ_AXIS_COUNT];
+static FAST_RAM_ZERO_INIT float setpointChangePerIteration[XYZ_AXIS_COUNT];
+static FAST_RAM_ZERO_INIT float deflectionChangePerIteration[XYZ_AXIS_COUNT];
+static FAST_RAM_ZERO_INIT float setpointReservoir[XYZ_AXIS_COUNT];
+static FAST_RAM_ZERO_INIT float deflectionReservoir[XYZ_AXIS_COUNT];
+
+// Configuration
+static FAST_RAM_ZERO_INIT float ffMaxRateScale;
+static FAST_RAM_ZERO_INIT float ffThumbLimit;
+static FAST_RAM_ZERO_INIT float ffMinSpread;
+
+void interpolatedSpInit(const pidProfile_t *pidProfile) {
+    ffMaxRateScale = pidProfile->ff_max_rate * 0.01f;
+    ffThumbLimit = pidProfile->ff_thumb_limit * 0.0001f;
+    ffMinSpread = pidProfile->ff_min_spread;
+}
+
+float interpolatedSpApply(int axis, float pidFrequency) {
```

Comment:
🤔 Yes, and filter based RC smoothing is still beneficial for other components of the PID loop. We only do this for the F component to alleviate the problems mentioned in the description. Let me think about it, but technically if we could both filter and interpolate when processing the RX signal we could get away with the setpoint interpolation part. I think that'd simplify this function a lot.
We'd still need the deflection and apply curve compensation on it though, it's being used to limit F contribution in `applyFFLimit()`

Reply:
This function is not directly related to RC smoothing.  It is almost equivalent to applying filter at RC update rate (RC interpolation/filtering is at PID rate).

The curve compensation can be probably simplified to, using filtered deflection as input and just calculate corresponding setpoint difference (using linear approximation around deflection). It will change the result slightly, but the difference shall be minimal.